packages/wrangler/src/api/startDevWorker/utils.ts

```diff
@@ -0,0 +1,43 @@
+import assert from "node:assert";
+
+export type MaybePromise<T> = T | Promise<T>;
+export type DeferredPromise<T> = {
+	promise: Promise<T>;
+	resolve: (_: MaybePromise<T>) => void;
+	reject: (_: Error) => void;
+};
+
+export function createDeferred<T>(
+	previousDeferred?: DeferredPromise<T>
+): DeferredPromise<T> {
+	let resolve, reject;
+	const newPromise = new Promise<T>((_resolve, _reject) => {
+		resolve = _resolve;
+		reject = _reject;
+	});
+	assert(resolve);
+	assert(reject);
+
+	// if passed a previousDeferred, ensure it is resolved with the newDeferred
+	// so that await-ers of previousDeferred are now await-ing newDeferred
+	previousDeferred?.resolve(newPromise);
```

Comment:
I'm a bit unclear on the use-case for chaining deferreds. If a deferred gets inherited, any holder of the previous deferred loses control over its lifecycle (the resolvers become inert since its promise has now already been resolved). Similarly, if the previous deferred has already been resolved before it gets inherited, the new deferred isn't allowed to extend the lifecycle of the previous one.

So whether or not a holder of a deferred maintains control over its lifecycle, and whether or not its lifecycle has been extended is unreliable. This seems error-prone and like it would lead to some insidious bugs.

Reply:
> If a deferred gets inherited, any holder of the previous deferred loses control over its lifecycle (the resolvers become inert since its promise has now already been resolved)

True

> Similarly, if the previous deferred has already been resolved before it gets inherited, the new deferred isn't allowed to extend the lifecycle of the previous one.

True

> So whether or not a holder of a deferred maintains control over its lifecycle, and whether or not its lifecycle has been extended is unreliable. This seems error-prone and like it would lead to some insidious bugs.

The point is not to extend the lifecycle of the old promise. It's to ensure the old promise is not left uncompleted. If the old promise is already completed (resolved/rejected), this line has no effect